### PR TITLE
ParticleSystem rendering in UI space

### PR DESCRIPTION
**Pull request by mvaligursky**
ParticleSystem component can be used on Entity inside hierarchy with parent being Screen component, allowing particles getting rendered integrated in UI layer (under button, on top of button ..)

Instructions:

* Add entity with ParticleSystem into hierarchy with root Screen. Note that it cannot be added as a child of screen, and its parent needs to be one of the Element components (Image, Group ..)
* Tick on 'ScreenSpace' option on the ParticleSystem
* Make sure the Layer is set to UI (same layer as the Screen) - otherwise sorting does not work

Engine example: engine/examples/#user-interface/button-particle.html

Supersedes #2178

### CODE REVIEW

File: src/framework/components/particle-system/component.js [hunk 1]
```diff
@@ -108,7 +109,7 @@ var depthLayer;
  * @property {boolean} depthWrite If enabled, the particles will write to the depth buffer. If disabled, the depth buffer is left unchanged and particles will be guaranteed to overwrite one another in the order in which they are rendered.
  * @property {boolean} noFog Disable fogging.
  * @property {boolean} localSpace Binds particles to emitter transformation rather then world space.
- * @property {number} numParticles Maximum number of simulated particles.
+ * @property {boolean} screenSpace Renders particles in 2D screen space. This needs to be set when particle system is part of hierarchy with {@link pc.ScreenComponent} at its root, and allows particle system to integrate with the rendering of {@link pc.ElementComponent}s. Note that an entity with ParticleSystem component cannot be parented directly to {@link pc.ScreenComponent}, but has to be a child of a {@link pc.ElementComponent}, for example {@link pc.LayoutGroupComponent}. * @property {number} numParticles Maximum number of simulated particles.
```
Comment (vkalpias) [inline src/framework/components/particle-system/component.js:112]: Looks like numParticles was deleted

Comment (mvaligursky) [reply src/framework/components/particle-system/component.js]: fixed

File: src/framework/components/particle-system/component.js [hunk 2]
```diff
@@ -108,6 +109,7 @@ var depthLayer;
  * @property {boolean} depthWrite If enabled, the particles will write to the depth buffer. If disabled, the depth buffer is left unchanged and particles will be guaranteed to overwrite one another in the order in which they are rendered.
  * @property {boolean} noFog Disable fogging.
  * @property {boolean} localSpace Binds particles to emitter transformation rather then world space.
+ * @property {boolean} screenSpace Renders particles in 2D screen space. This needs to be set when particle system is part of hierarchy with {@link pc.ScreenComponent} at its root, and allows particle system to integrate with the rendering of {@link pc.ElementComponent}s. Note that an entity with ParticleSystem component cannot be parented directly to {@link pc.ScreenComponent}, but has to be a child of a {@link pc.ElementComponent}, for example {@link pc.LayoutGroupComponent}.
```
Comment (willeastcott) [inline src/framework/components/particle-system/component.js:112]: The screen doesn't have to be at the root, right? It just has to be an ancestor in the hierarchy, right?

Comment (willeastcott) [reply src/framework/components/particle-system/component.js:112]: Also, what's the motivation for having the `screenSpace` property anyway? If the particlesystem component is a descendant of an entity with a screen component, shouldn't screen space be assumed automatically?

Comment (mvaligursky) [reply src/framework/components/particle-system/component.js]: I changed the doc to use "ancestor", thanks.

Comment (mvaligursky) [reply src/framework/components/particle-system/component.js]: > Also, what's the motivation for having the `screenSpace` property anyway? If the particlesystem component is a descendant of an entity with a screen component, shouldn't screen space be assumed automatically?

The motivation is avoiding the complexity / performance impact of that. If a child entity is added to a parent, the code would need to check if any of the descendant entities contain ParticleSystem and if there is a Screen component ascendant. setParent could get expensive.

**Pull request merged**